src/main/java/com/musala/atmosphere/client/websocket/ClientDispatcher.java

```diff
@@ -0,0 +1,308 @@
+        RequestMessage requestMessage = new RequestMessage(MessageAction.ROUTING_ACTION, action, args);
+        requestMessage.setDeviceId(deviceId);
+        // The session identifier must be unique not only per WebSocket Session but and per action in case in multiple
+        // threads scenario. Otherwise the WebSocketCommunicationManager can return a response from another action.
+        String sessionIdentifier = this.session.getId() + "_" + action.toString();
+        requestMessage.setSessionId(sessionIdentifier);
+        requestMessage.setPasskey(invocationPasskey);
+
+        return requestMessage;
+    }
+
+    /**
+     * TODO: Consider to migrate this wait logic on the Server.
+     *
+     * Gets a {@link DeviceAllocationInformation} instance with the given {@link DeviceSelector device characteristics}.
+     *
+     * @param deviceSelector
+     *        - required {@link DeviceSelector parameters} needed to construct new {@link DeviceAllocationInformation}
+     *        instance.
+     * @return a {@link DeviceAllocationInformation} instance with the given device selector.
+     * @throws NoAvailableDeviceFoundException
+     *         - when there is no available device
+     */
+    public DeviceAllocationInformation getDeviceDescriptor(DeviceSelector deviceSelector,
+                                                           int allocateDeviceRetryCount) {
+        while (allocateDeviceRetryCount > 0) {
+            try {
+                RequestMessage request = new RequestMessage(MessageAction.DEVICE_ALLOCATION_INFORMATION,
+                                                            deviceSelector);
+                request.setSessionId(session.getId());
+
+                DeviceAllocationInformation allocationInfo = null;
+                ResponseMessage response = sendRequest(request, session);
+
+                if (response.getMessageAction() == MessageAction.ERROR) {
+                    if (response.getException() instanceof NoDeviceMatchingTheGivenSelectorException) {
+                        throw (NoDeviceMatchingTheGivenSelectorException) response.getException();
+                    } else if (response.getException() instanceof NoAvailableDeviceFoundException) {
+                        throw (NoAvailableDeviceFoundException) response.getException();
+                    }
+                }
+                allocationInfo = (DeviceAllocationInformation) response.getData();
+
+                return allocationInfo;
+            } catch (NoAvailableDeviceFoundException e) {
+                try {
+                    Thread.sleep(1000);
+                } catch (InterruptedException e1) {
+                    // Nothing to do here.
+                }
+                allocateDeviceRetryCount--;
+            } catch (NoDeviceMatchingTheGivenSelectorException e) {
+                break;
+            }
```

Comment:
What you are doing here is called expection handling. I could not find very good reference for this but for example the second section here: https://glenndejaeger.wordpress.com/2010/04/05/programming-antipatterns/. If you throw exceptions to force a "Go-to" like instructions this is definitely bad way to structure your program. Please revise this program segment